File: November/Day29/380_Insert_Delete_GetRandom_O(1).cpp
class RandomizedSet {
public:
    unordered_map<int,int>mp;
    vector<int>arr;
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
     if(mp.count(val))
     {
         return false;
     }
     arr.push_back(val);
     mp[val] = arr.size()-1;
     return true;
    }
    
    bool remove(int val) {
     if(mp.count(val))
     {
     int index = mp[val];
     int v = arr.back();
     arr[index] = v;
     mp[v] = index;
     arr.pop_back();
     mp.erase(val);
    return true;
     }
    return false;
    }
    
    int getRandom() {
      int index = rand()%arr.size();
      return arr[index];  
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: November/Day29/380_Insert_Delete_GetRandom_O(1).java
class RandomizedSet {
    List<Integer> ls = new ArrayList<>();
    HashMap<Integer,Integer> map = new HashMap<>();
    public RandomizedSet() {
        
    }
    
    public boolean insert(int val) {
        if(map.containsKey(val))return false;
        map.put(val,ls.size());
        ls.add(val);
        return true;
    }
    
    public boolean remove(int val) {
        if(!map.containsKey(val))return false;
        
        int index= map.get(val);
        int lastel= ls.get(ls.size()-1);
        ls.set(index,lastel);
        map.put(lastel,index);
        map.remove(val);
        ls.remove(ls.size()-1);
        return true;
    }
    
    public int getRandom() {
        int randomIdx = (int) (Math.random() * ls.size());
        return ls.get(randomIdx);
        
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */
